packages/loader/container-loader/src/container.ts

```diff
@@ -216,24 +213,24 @@ class LifecycleStateHandler {
      * Transitions the container's lifecycle state
      */
     public changeState(newState: LifecycleState) {
-        // TODO: Figure our the possible state transitions...
+        if (logIfFalse(this.isStateChangeAllowed(newState), this.logger, "InvalidStateChange")) {
+            this._state = newState;
+        }
+    }
+
+    constructor(private readonly logger: ITelemetryLogger) {}
+
+    private isStateChangeAllowed(newState: LifecycleState): boolean {
         switch (newState) {
-            case LifecycleState.Loaded:
-                assert(this._state !== LifecycleState.Created,
-                    0x27e /* "Must go through loading state before loaded" */);
-
-                // It's conceivable the container could be closed when this is called
-                // Only transition states if currently loading
-                if (this._state === LifecycleState.Loading) {
-                    this._state = newState;
-                }
-                break;
-            default:
-                this._state = newState;
+            case LifecycleState.Created: return false;
+            case LifecycleState.Loading: return this._state === LifecycleState.Created;
+            case LifecycleState.Loaded: return this._state === LifecycleState.Loading;
```

Comment:
This should switch back to your previous iteration -- the way the code was before.